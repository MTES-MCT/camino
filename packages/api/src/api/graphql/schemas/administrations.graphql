# import * from 'metas.graphql'
# import * from 'administrations.graphql'
# import * from 'utilisateurs.graphql'
# import * from 'titres.graphql'
# import * from 'metas.graphql'
# import * from 'metas-activites.graphql'
# import * from 'territoires.graphql'

type Administration {
  id: ID!
  nom: String
  typeId: String
  departementId: String
  regionId: String
  abreviation: String
  service: String
  url: String
  email: String
  telephone: String
  adresse1: String
  adresse2: String
  codePostal: String
  commune: String
  cedex: String
  utilisateurs: [Utilisateur]
  emailsModification: Boolean
  gestionnaireTitres: [Titre]
  localeTitres: [Titre]
  titresTypesTitresStatuts: [TitreTypeTitreStatut]
  titresTypesEtapesTypes: [TitreTypeEtapeType]
  activitesTypes: [ActiviteType]
  modification: Boolean
}

type ActiviteTypeEmail {
  activiteTypeId: String
  email: String
}

input InputAdministrationTitreTypeTitreStatut {
  administrationId: ID!
  titreTypeId: ID!
  titreStatutId: ID!
  titresModificationInterdit: Boolean
  demarchesModificationInterdit: Boolean
  etapesModificationInterdit: Boolean
}

input InputAdministrationTitreTypeEtapeType {
  administrationId: ID!
  titreTypeId: ID!
  etapeTypeId: ID!
  lectureInterdit: Boolean
  modificationInterdit: Boolean
  creationInterdit: Boolean
}

input InputAdministrationActiviteType {
  administrationId: ID!
  activiteTypeId: ID!
  lectureInterdit: Boolean
  modificationInterdit: Boolean
}

input InputAdministrationActiviteTypeEmail {
  administrationId: ID!
  activiteTypeId: ID!
  email: String
}
