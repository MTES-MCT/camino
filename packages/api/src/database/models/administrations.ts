import { Model } from 'objection'

import { IAdministration } from '../../types'

import AdministrationsTitresTypesEtapesTypes from './administrations-titres-types-etapes-types'
import AdministrationsTitresTypesTitresStatuts from './administrations-titres-types-titres-statuts'
import Utilisateurs from './utilisateurs'
import Titres from './titres'
import ActivitesTypes from './activites-types'

interface Administrations extends IAdministration {}

class Administrations extends Model {
  public static tableName = 'administrations'

  public static jsonSchema = {
    type: 'object',
    required: ['id'],

    properties: {
      id: { type: 'string', maxLength: 64 }
    }
  }

  static relationMappings = () => ({
    titresTypesTitresStatuts: {
      relation: Model.HasManyRelation,
      modelClass: AdministrationsTitresTypesTitresStatuts,
      join: {
        from: 'administrations.id',
        to: 'administrations__titresTypes__titresStatuts.administrationId'
      }
    },

    titresTypesEtapesTypes: {
      relation: Model.HasManyRelation,
      modelClass: AdministrationsTitresTypesEtapesTypes,
      join: {
        from: 'administrations.id',
        to: 'administrations__titresTypes__etapesTypes.administrationId'
      }
    },

    utilisateurs: {
      relation: Model.HasManyRelation,
      modelClass: Utilisateurs,
      join: {
        from: 'administrations.id',
        to: 'utilisateurs.administrationId'
      }
    },

    gestionnaireTitres: {
      relation: Model.ManyToManyRelation,
      modelClass: Titres,
      join: {
        from: 'administrations.id',
        through: {
          from: 'titresAdministrationsGestionnaires.administrationId',
          to: 'titresAdministrationsGestionnaires.titreId'
        },
        to: 'titres.id'
      }
    },

    activitesTypes: {
      relation: Model.ManyToManyRelation,
      modelClass: ActivitesTypes,
      join: {
        from: 'administrations.id',
        through: {
          from: 'administrations__activitesTypes.administrationId',
          to: 'administrations__activitesTypes.activiteTypeId',
          extra: ['modificationInterdit', 'lectureInterdit']
        },
        to: 'activitesTypes.id'
      }
    },

    activitesTypesEmails: {
      relation: Model.ManyToManyRelation,
      modelClass: ActivitesTypes,
      join: {
        from: 'administrations.id',
        through: {
          from: 'administrations__activitesTypes__emails.administrationId',
          to: 'administrations__activitesTypes__emails.activiteTypeId',
          extra: ['email']
        },
        to: 'activitesTypes.id'
      }
    }
  })
}

export default Administrations
