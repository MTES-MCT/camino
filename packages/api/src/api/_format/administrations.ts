import { IAdministration } from '../../types'

import { titresFormat } from './titres'
import { utilisateurFormat } from './utilisateurs'
import {
  Administrations,
  ADMINISTRATION_TYPES
} from 'camino-common/src/administrations'
import { Regions } from 'camino-common/src/region'
import { Departements } from 'camino-common/src/departement'

/**
 * Formate une administration en fonction du profil de l'utilisateur
 *
 * @param administration - Administration à formater
 * @returns Une administration formatée
 *
 */

const administrationFormat = (administration: IAdministration) => {
  administration.gestionnaireTitres =
    administration.gestionnaireTitres &&
    titresFormat(administration.gestionnaireTitres)

  administration.localeTitres =
    administration.localeTitres && titresFormat(administration.localeTitres)

  administration.utilisateurs =
    administration.utilisateurs?.map(utilisateurFormat)

  const adminis = Administrations[administration.id]
  administration.type = ADMINISTRATION_TYPES[adminis.typeId]

  for (const key of Object.keys(adminis)) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    administration[key] = adminis[key]
  }

  if (adminis.departementId) {
    administration.departement = Departements[adminis.departementId]
  }
  if (adminis.regionId) {
    administration.region = Regions[adminis.regionId]
  }

  return administration
}

export { administrationFormat }
