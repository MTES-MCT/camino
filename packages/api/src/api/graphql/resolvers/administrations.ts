import { GraphQLResolveInfo } from 'graphql'

import { IAdministrationActiviteType, IAdministrationTitreTypeEtapeType, IAdministrationTitreTypeTitreStatut, IAdministrationActiviteTypeEmail, Context } from '../../../types.js'

import {
  administrationGet,
  administrationsGet,
  administrationTitreTypeTitreStatutUpsert,
  administrationTitreTypeTitreStatutDelete,
  administrationTitreTypeEtapeTypeDelete,
  administrationTitreTypeEtapeTypeUpsert,
  administrationActiviteTypeDelete,
  administrationActiviteTypeUpsert,
  administrationActiviteTypeEmailCreate,
  administrationActiviteTypeEmailDelete,
} from '../../../database/queries/administrations.js'

import { fieldsBuild } from './_fields-build.js'

import { administrationFormat } from '../../_format/administrations.js'
import { emailCheck } from '../../../tools/email-check.js'
import { isSuper } from 'camino-common/src/roles.js'
import { canReadActivitesTypesEmails, canReadAdministrations, canEditEmails } from 'camino-common/src/permissions/administrations.js'
import administrationsActivitesTypesEmails from '../../../database/models/administrations-activites-types-emails.js'
import { isAdministrationId } from 'camino-common/src/static/administrations.js'

const administration = async ({ id }: { id: string }, { user }: Context, info: GraphQLResolveInfo) => {
  try {
    if (!canReadAdministrations(user)) {
      return null
    }
    const fields = fieldsBuild(info)

    const administration = await administrationGet(id, { fields }, user)

    if (!administration) {
      return null
    }

    return administrationFormat(administration)
  } catch (e) {
    console.error(e)

    throw e
  }
}

export const administrationActivitesTypesEmails = async ({ id }: { id: string }, { user }: Context): Promise<IAdministrationActiviteTypeEmail[]> => {
  try {
    if (isAdministrationId(id)) {
      if (!canReadActivitesTypesEmails(user, id)) {
        throw new Error('droit insuffisant')
      }

      return administrationsActivitesTypesEmails.query().where('administrationId', id)
    } else {
      throw new Error('l’identifiant est inconnu')
    }
  } catch (e) {
    console.error(e)

    throw e
  }
}

const administrations = async (_: unknown, { user }: Context, info: GraphQLResolveInfo) => {
  try {
    if (!canReadAdministrations(user)) {
      return []
    }
    const fields = fieldsBuild(info)

    const administrations = await administrationsGet({ fields }, user)

    if (!administrations.length) return []

    return administrations.map(administrationFormat)
  } catch (e) {
    console.error(e)

    throw e
  }
}

const administrationTitreTypeTitreStatutModifier = async (
  {
    administrationTitreTypeTitreStatut,
  }: {
    administrationTitreTypeTitreStatut: IAdministrationTitreTypeTitreStatut
  },
  { user }: Context,
  info: GraphQLResolveInfo
) => {
  try {
    if (!isSuper(user)) {
      throw new Error('droits insuffisants')
    }

    const fields = fieldsBuild(info)

    if (
      !administrationTitreTypeTitreStatut.titresModificationInterdit &&
      !administrationTitreTypeTitreStatut.demarchesModificationInterdit &&
      !administrationTitreTypeTitreStatut.etapesModificationInterdit
    ) {
      await administrationTitreTypeTitreStatutDelete(
        administrationTitreTypeTitreStatut.administrationId,
        administrationTitreTypeTitreStatut.titreTypeId,
        administrationTitreTypeTitreStatut.titreStatutId
      )
    } else {
      await administrationTitreTypeTitreStatutUpsert(administrationTitreTypeTitreStatut)
    }

    return await administrationGet(administrationTitreTypeTitreStatut.administrationId, { fields }, user)
  } catch (e) {
    console.error(e)

    throw e
  }
}

const administrationTitreTypeEtapeTypeModifier = async (
  {
    administrationTitreTypeEtapeType,
  }: {
    administrationTitreTypeEtapeType: IAdministrationTitreTypeEtapeType
  },
  { user }: Context,
  info: GraphQLResolveInfo
) => {
  try {
    if (!isSuper(user)) {
      throw new Error('droits insuffisants')
    }

    const fields = fieldsBuild(info)

    if (!administrationTitreTypeEtapeType.lectureInterdit && !administrationTitreTypeEtapeType.modificationInterdit && !administrationTitreTypeEtapeType.creationInterdit) {
      await administrationTitreTypeEtapeTypeDelete(administrationTitreTypeEtapeType.administrationId, administrationTitreTypeEtapeType.titreTypeId, administrationTitreTypeEtapeType.etapeTypeId)
    } else {
      await administrationTitreTypeEtapeTypeUpsert(administrationTitreTypeEtapeType)
    }

    return await administrationGet(administrationTitreTypeEtapeType.administrationId, { fields }, user)
  } catch (e) {
    console.error(e)

    throw e
  }
}

const administrationActiviteTypeModifier = async ({ administrationActiviteType }: { administrationActiviteType: IAdministrationActiviteType }, { user }: Context, info: GraphQLResolveInfo) => {
  try {
    if (!isSuper(user)) {
      throw new Error('droits insuffisants')
    }

    const fields = fieldsBuild(info)

    if (!administrationActiviteType.lectureInterdit && !administrationActiviteType.modificationInterdit) {
      await administrationActiviteTypeDelete(administrationActiviteType.administrationId, administrationActiviteType.activiteTypeId)
    } else {
      await administrationActiviteTypeUpsert(administrationActiviteType)
    }

    return await administrationGet(administrationActiviteType.administrationId, { fields }, user)
  } catch (e) {
    console.error(e)

    throw e
  }
}

const administrationActiviteTypeEmailCreer = async (
  { administrationActiviteTypeEmail }: { administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail },
  { user }: Context,
  info: GraphQLResolveInfo
) => {
  try {
    const administration = await administrationGet(administrationActiviteTypeEmail.administrationId, { fields: { id: {} } }, user)

    if (!administration || !canEditEmails(user, administration.id)) {
      throw new Error('droits insuffisants')
    }

    const fields = fieldsBuild(info)
    const email = administrationActiviteTypeEmail.email?.toLowerCase()
    if (!email || !emailCheck(email)) throw new Error('email invalide')

    await administrationActiviteTypeEmailCreate({
      ...administrationActiviteTypeEmail,
      email,
    })

    return await administrationGet(administrationActiviteTypeEmail.administrationId, { fields }, user)
  } catch (e) {
    console.error(e)

    // Un doublon d'email + admin ID + type d'activité génère une erreur spécifique.
    if ((e as Error).name === 'UniqueViolationError') {
      throw new Error('cette notification est déjà prise en compte')
    }

    throw e
  }
}

const administrationActiviteTypeEmailSupprimer = async (
  { administrationActiviteTypeEmail }: { administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail },
  { user }: Context,
  info: GraphQLResolveInfo
) => {
  try {
    const administration = await administrationGet(administrationActiviteTypeEmail.administrationId, { fields: { id: {} } }, user)

    if (!administration || !canEditEmails(user, administration.id)) {
      throw new Error('droits insuffisants')
    }

    const fields = fieldsBuild(info)
    const email = administrationActiviteTypeEmail.email?.toLowerCase()
    if (!email || !emailCheck(email)) throw new Error('email invalide')

    await administrationActiviteTypeEmailDelete({
      ...administrationActiviteTypeEmail,
      email,
    })

    return await administrationGet(administrationActiviteTypeEmail.administrationId, { fields }, user)
  } catch (e) {
    console.error(e)

    throw e
  }
}

export {
  administration,
  administrations,
  administrationTitreTypeTitreStatutModifier,
  administrationTitreTypeEtapeTypeModifier,
  administrationActiviteTypeModifier,
  administrationActiviteTypeEmailCreer,
  administrationActiviteTypeEmailSupprimer,
}
