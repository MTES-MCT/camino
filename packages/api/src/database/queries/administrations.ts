import {
  IAdministrationActiviteType,
  IAdministrationActiviteTypeEmail,
  IAdministrationTitreTypeEtapeType,
  IAdministrationTitreTypeTitreStatut,
  IFields,
  IUtilisateur
} from '../../types.js'

import graphBuild from './graph/build.js'
import { fieldsFormat } from './graph/fields-format.js'
import options from './_options.js'

import Administrations from '../models/administrations.js'
import { administrationsQueryModify } from './permissions/administrations.js'
import AdministrationsTitresTypesTitresStatuts from '../models/administrations-titres-types-titres-statuts.js'
import AdministrationsTitresTypesEtapesTypes from '../models/administrations-titres-types-etapes-types.js'
import AdministrationsActivitesTypes from '../models/administrations-activites-types.js'
import AdministrationsActivitesTypesEmails from '../models/administrations-activites-types-emails.js'

const administrationsQueryBuild = (
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  const graph = fields
    ? graphBuild(fields, 'administrations', fieldsFormat)
    : options.administrations.graph

  const q = Administrations.query().withGraphFetched(graph)

  administrationsQueryModify(q, user)

  return q
}

const administrationGet = async (
  id: string,
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  const q = administrationsQueryBuild({ fields }, user)

  return q.findById(id)
}

const administrationsGet = async (
  { fields }: { fields?: IFields },
  user: IUtilisateur | null | undefined
) => {
  return administrationsQueryBuild({ fields }, user)
}

const administrationTitreTypeTitreStatutUpsert = async (
  administrationTitreTypeTitreStatut: IAdministrationTitreTypeTitreStatut
) =>
  AdministrationsTitresTypesTitresStatuts.query().upsertGraph(
    administrationTitreTypeTitreStatut,
    { insertMissing: true }
  )

const administrationTitreTypeTitreStatutDelete = async (
  administrationId: string,
  titreTypeId: string,
  statutTypeId: string
) =>
  AdministrationsTitresTypesTitresStatuts.query().deleteById([
    administrationId,
    titreTypeId,
    statutTypeId
  ])

const administrationTitreTypeEtapeTypeUpsert = async (
  administrationTitreTypeEtapeType: IAdministrationTitreTypeEtapeType
) =>
  AdministrationsTitresTypesEtapesTypes.query().upsertGraph(
    administrationTitreTypeEtapeType,
    { insertMissing: true }
  )

const administrationTitreTypeEtapeTypeDelete = async (
  administrationId: string,
  titreTypeId: string,
  etapeTypeId: string
) =>
  AdministrationsTitresTypesEtapesTypes.query().deleteById([
    administrationId,
    titreTypeId,
    etapeTypeId
  ])

const administrationActiviteTypeUpsert = async (
  administrationActiviteType: IAdministrationActiviteType
) =>
  AdministrationsActivitesTypes.query().upsertGraph(
    administrationActiviteType,
    {
      insertMissing: true
    }
  )

const administrationActiviteTypeDelete = async (
  administrationId: string,
  ActiviteTypeId: string
) =>
  AdministrationsActivitesTypes.query().deleteById([
    administrationId,
    ActiviteTypeId
  ])

const administrationActiviteTypeEmailCreate = async (
  administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail
) =>
  AdministrationsActivitesTypesEmails.query().insertGraph(
    administrationActiviteTypeEmail
  )

const administrationActiviteTypeEmailDelete = async (
  administrationActiviteTypeEmail: IAdministrationActiviteTypeEmail
) => {
  const { activiteTypeId, administrationId, email } =
    administrationActiviteTypeEmail

  return AdministrationsActivitesTypesEmails.query().delete().where({
    activite_type_id: activiteTypeId,
    administration_id: administrationId,
    email
  })
}

export {
  administrationGet,
  administrationsGet,
  administrationTitreTypeTitreStatutUpsert,
  administrationTitreTypeTitreStatutDelete,
  administrationTitreTypeEtapeTypeUpsert,
  administrationTitreTypeEtapeTypeDelete,
  administrationActiviteTypeUpsert,
  administrationActiviteTypeDelete,
  administrationActiviteTypeEmailCreate,
  administrationActiviteTypeEmailDelete
}
