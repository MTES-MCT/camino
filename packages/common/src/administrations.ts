import { Administration, AdministrationId, sortedAdministrations } from './static/administrations.js'
import { CommuneId } from './static/communes.js'
import { DepartementId, Departements, toDepartementId } from './static/departement.js'
import { SecteursMaritimes, getDepartementsBySecteurs } from './static/facades.js'
import { onlyUnique, isNullOrUndefined } from './typescript-tools.js'

// calcule tous les départements d'une étape
const titreEtapeAdministrationsDepartementsBuild = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): DepartementId[] => {
  if (!communes) {
    throw new Error('les communes ne sont pas chargées')
  }

  const departements = communes.map(id => toDepartementId(id))

  if (secteursMaritimes) {
    departements.push(...getDepartementsBySecteurs(secteursMaritimes))
  }

  return departements.filter(onlyUnique)
}

export const getAdministrationsLocales = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): AdministrationId[] => {
  const titreDepartementsIds = titreEtapeAdministrationsDepartementsBuild(communes, secteursMaritimes)
  const titreRegionsIds = titreDepartementsIds.map(departementId => Departements[departementId].regionId).filter(onlyUnique)

  // calcule toutes les administrations qui couvrent ces départements et régions
  return sortedAdministrations.reduce((titreEtapeAdministrations: AdministrationId[], administration: Administration) => {
    // si le département ou la région de l'administration ne fait pas partie de ceux de l'étape
    const isAdministrationLocale =
      (administration.departementId && titreDepartementsIds.includes(administration.departementId)) || (administration.regionId && titreRegionsIds.includes(administration.regionId))
    // alors l'administration n'est pas rattachée à l'étape
    if (isNullOrUndefined(isAdministrationLocale) || !isAdministrationLocale) return titreEtapeAdministrations

    titreEtapeAdministrations.push(administration.id)

    return titreEtapeAdministrations
  }, [])
}
