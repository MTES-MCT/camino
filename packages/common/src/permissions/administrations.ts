import { isAdministration, isAdministrationAdmin, isAdministrationEditeur, isSuper, User } from '../roles.js'
import { AdministrationId, Administrations, sortedAdministrations } from '../static/administrations.js'
import { Departements } from '../static/departement.js'

export const canReadActivitesTypesEmails = (user: User, administrationId: AdministrationId) => {
  if (!canReadAdministrations(user)) {
    return false
  }

  if (isSuper(user) || (isAdministration(user) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministration(user)) {
    const administration = Administrations[user.administrationId]

    const administrationIds: AdministrationId[] = []

    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      administrationIds.push(...sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id))
    }

    administrationIds.push(user.administrationId)

    return administrationIds.includes(administrationId)
  }

  return false
}

export const canReadAdministrations = (user: User) => isSuper(user) || isAdministration(user)

export const canEditEmails = (user: User, administrationId: AdministrationId): boolean => {
  if (isSuper(user) || ((isAdministrationAdmin(user) || isAdministrationEditeur(user)) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    // Membre d'une DREAL/DEAL vis-à-vis de la DREAL elle-même,
    // ou d'un DREAL/DEAL vis-à-vis d'une administration qui dépend d'elles
    // Admin ou éditeur : modifications
    // Admin, éditeur ou lecteur : lecture

    const administration = Administrations[user.administrationId]
    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      const administrationIds = sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id)

      if (administration.typeId === 'dre' || administration.typeId === 'dea') {
        administrationIds.push(user.administrationId)
      }

      return administrationIds.includes(administrationId)
    }
  }

  return false
}
