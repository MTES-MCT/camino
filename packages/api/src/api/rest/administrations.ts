import { Request as JWTRequest } from 'express-jwt'
import { HTTP_STATUS } from 'camino-common/src/http'

import { CustomResponse } from './express-type'
import { AdminUserNotNull, User, UserNotNull } from 'camino-common/src/roles'
import { Pool } from 'pg'
import { AdministrationId, administrationIdValidator } from 'camino-common/src/static/administrations'
import { canReadAdministrations } from 'camino-common/src/permissions/administrations'
import { deleteAdministrationActiviteTypeEmail, getActiviteTypeEmailsByAdministrationId, getUtilisateursByAdministrationId, insertAdministrationActiviteTypeEmail } from './administrations.queries'
import { AdministrationActiviteTypeEmail } from 'camino-common/src/administrations'
import { CaminoApiError } from '../../types'
import { ZodUnparseable } from '../../tools/fp-tools'
import { DeepReadonly } from 'camino-common/src/typescript-tools'
import { DbQueryAccessError } from '../../pg-database'
import { Effect, Match } from 'effect'
import { RestNewPostCall } from '../../server/rest'

export const getAdministrationUtilisateurs = (pool: Pool) => async (req: JWTRequest<User>, res: CustomResponse<AdminUserNotNull[]>) => {
  const user = req.auth

  const parsed = administrationIdValidator.safeParse(req.params.administrationId)

  if (!parsed.success) {
    console.warn(`l'administrationId est obligatoire`)
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else if (!canReadAdministrations(user)) {
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else {
    try {
      res.json(await getUtilisateursByAdministrationId(pool, parsed.data))
    } catch (e) {
      console.error(e)

      res.sendStatus(HTTP_STATUS.INTERNAL_SERVER_ERROR)
    }
  }
}

export const getAdministrationActiviteTypeEmails = (pool: Pool) => async (req: JWTRequest<User>, res: CustomResponse<AdministrationActiviteTypeEmail[]>) => {
  const user = req.auth

  const parsed = administrationIdValidator.safeParse(req.params.administrationId)

  if (!parsed.success) {
    console.warn(`l'administrationId est obligatoire`)
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else if (!canReadAdministrations(user)) {
    res.sendStatus(HTTP_STATUS.FORBIDDEN)
  } else {
    try {
      res.json(await getActiviteTypeEmailsByAdministrationId(pool, parsed.data))
    } catch (e) {
      console.error(e)

      res.sendStatus(HTTP_STATUS.INTERNAL_SERVER_ERROR)
    }
  }
}

export const addAdministrationActiviteTypeEmails: RestNewPostCall<'/rest/administrations/:administrationId/activiteTypeEmails'> = (
  pool: Pool,
  user: DeepReadonly<UserNotNull>,
  body: DeepReadonly<AdministrationActiviteTypeEmail>,
  params: { administrationId: AdministrationId }
): Effect.Effect<boolean, CaminoApiError<'Accès interdit' | ZodUnparseable | DbQueryAccessError>> => {
  return Effect.Do.pipe(
    Effect.filterOrFail(
      () => canReadAdministrations(user),
      () => ({ message: 'Accès interdit' as const })
    ),
    Effect.flatMap(() => insertAdministrationActiviteTypeEmail(pool, params.administrationId, body)),
    Effect.mapError(caminoError =>
      Match.value(caminoError.message).pipe(
        Match.when('Accès interdit', () => ({ ...caminoError, status: HTTP_STATUS.FORBIDDEN })),
        Match.when('Problème de validation de données', () => ({ ...caminoError, status: HTTP_STATUS.BAD_REQUEST })),
        Match.when("Impossible d'accéder à la base de données", () => ({ ...caminoError, status: HTTP_STATUS.INTERNAL_SERVER_ERROR })),
        Match.exhaustive
      )
    )
  )
}

export const deleteAdministrationActiviteTypeEmails: RestNewPostCall<'/rest/administrations/:administrationId/activiteTypeEmails/delete'> = (
  pool: Pool,
  user: DeepReadonly<UserNotNull>,
  body: DeepReadonly<AdministrationActiviteTypeEmail>,
  params: { administrationId: AdministrationId }
): Effect.Effect<boolean, CaminoApiError<'Accès interdit' | ZodUnparseable | DbQueryAccessError>> => {
  return Effect.Do.pipe(
    Effect.filterOrFail(
      () => canReadAdministrations(user),
      () => ({ message: 'Accès interdit' as const })
    ),
    Effect.flatMap(() => deleteAdministrationActiviteTypeEmail(pool, params.administrationId, body)),
    Effect.mapError(caminoError =>
      Match.value(caminoError.message).pipe(
        Match.when('Accès interdit', () => ({ ...caminoError, status: HTTP_STATUS.FORBIDDEN })),
        Match.when('Problème de validation de données', () => ({ ...caminoError, status: HTTP_STATUS.BAD_REQUEST })),
        Match.when("Impossible d'accéder à la base de données", () => ({ ...caminoError, status: HTTP_STATUS.INTERNAL_SERVER_ERROR })),
        Match.exhaustive
      )
    )
  )
}
