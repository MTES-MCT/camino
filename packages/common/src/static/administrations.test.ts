import { ADMINISTRATION_IDS, isAdministrationId, sortedAdministrationTypes } from './administrations.js'
import { test, expect } from 'vitest'

test('sortedAdministrationTypes', () => {
  expect(sortedAdministrationTypes).toMatchSnapshot()
})

test('isAdministrationId', () => {
  expect(isAdministrationId(null)).toBe(false)
  expect(isAdministrationId(undefined)).toBe(false)
  expect(isAdministrationId('notAnAdmin')).toBe(false)
  expect(isAdministrationId([ADMINISTRATION_IDS.BRGM])).toBe(false)

  Object.values(ADMINISTRATION_IDS).forEach(administrationId => expect(isAdministrationId(administrationId)).toBe(true))
})
