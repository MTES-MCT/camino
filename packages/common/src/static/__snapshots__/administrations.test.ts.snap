// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sortedAdministrationTypes 1`] = `
[
  {
    "id": "aut",
    "nom": "Autorité",
  },
  {
    "id": "dea",
    "nom": "Déal",
  },
  {
    "id": "dre",
    "nom": "Dréal",
  },
  {
    "id": "min",
    "nom": "Ministère",
  },
  {
    "id": "ope",
    "nom": "Opérateur",
  },
  {
    "id": "pre",
    "nom": "Préfecture",
  },
]
`;
